src/common/async/throttle.h

```diff
@@ -0,0 +1,274 @@
+ * Foundation. See file COPYING.
+ *
+ */
+
+#ifndef CEPH_ASYNC_THROTTLE_H
+#define CEPH_ASYNC_THROTTLE_H
+
+#include <mutex>
+#include <boost/intrusive/list.hpp>
+#include "completion.h"
+#include "common/perf_counters.h"
+
+namespace ceph::async {
+
+/**
+ * An asynchronous throttle implementation for use with boost::asio.
+ *
+ * Example use:
+ *
+ *   // process a list of jobs, running up to 10 jobs in parallel
+ *   boost::asio::io_context context;
+ *   ceph::async::Throttle throttle(context.get_executor(), 10, {});
+ *
+ *   for (auto& job : jobs) {
+ *     // request a throttle unit for this job
+ *     throttle.async_get(1, [&] (boost::system::error_code ec) {
+ *         if (!ec) { // throttle was granted
+ *           job.process();
+ *           throttle.put(1);
+ *         }
+ *       });
+ *   }
+ *   context.run();
+ */
+template <typename Executor1>
+class Throttle;
+
+
+namespace throttle_counters {
+
+enum {
+  l_first = 711262,
+  l_val,
+  l_max,
+  l_pending,
+  l_get,
+  l_get_sum,
+  l_put,
+  l_put_sum,
+  l_wait,
+  l_wait_sum,
+  l_cancel,
+  l_cancel_sum,
+  l_latency,
+  l_last,
+};
+
+PerfCountersRef build(CephContext *cct, const std::string& name);
+
+} // namespace throttle_counters
+
+namespace detail {
+
+// Throttle base contains everything that doesn't depend on the Executor type
+class BaseThrottle {
+ public:
+  BaseThrottle(size_t maximum, PerfCountersRef&& counters);
+  ~BaseThrottle();
+
+  /// returns a number of previously-granted throttle units, then attempts to
+  /// grant the next pending requests
+  void put(size_t count);
+
+  /// adjust the maximum throttle. if a previous call to async_set_maximum() is
+  /// still pending, its completion handler is invoked with operation_aborted.
+  /// if pending calls to async_get() can be satisfied by the new maximum
+  /// throttle, their completion handlers are invoked
+  void set_maximum(size_t count);
+
+  /// cancel any pending calls to async_get() and async_set_maximum(), invoking
+  /// their completion handlers with operation_aborted
+  void cancel();
+
+  PerfCounters* get_counters() const { return counters.get(); }
+
+ protected:
+  template <typename Executor1, typename Handler>
+  void start_get(const Executor1& ex1, size_t count, Handler&& handler);
+
+  template <typename Executor1, typename Handler>
+  void start_set_maximum(const Executor1& ex1, size_t count, Handler&& handler);
+
+ private:
+  std::mutex mutex;
+  size_t value = 0; //< amount of outstanding throttle
+  size_t maximum; //< maximum amount of throttle
+
+  PerfCountersRef counters;
+  using Clock = ceph::mono_clock;
+
+  struct GetRequest : public boost::intrusive::list_base_hook<> {
+    const size_t count; //< number of throttle units requested
+    const Clock::time_point started;
+    GetRequest(size_t count, Clock::time_point started) noexcept
+      : count(count), started(started) {}
+  };
+  using GetCompletion = Completion<void(boost::system::error_code),
```

Comment:
OoC, any reason to use boost::system::error_code and not std::error_code?

Reply:
Boost.Asio and Beast force that on us, unfortunately.

Reply:
one of the main reasons is that `boost::system::error_code` is part of the interface for stackful coroutines, which is how rgw is using this throttle: `request_throttle.async_get(1, yield[ec]);`

you can read more about error handling with stackful coroutines here: http://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/overview/core/spawn.html

at a higher level, if you're writing an 'asynchronous initiating function' like `async_get()` and want to support any kind of `CompletionToken` that follows the [extensible asynchronous model](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4045.pdf), your `Signature` has to have `boost::system::error_code` as the first parameter. asio and beast both do this throughout

i'd love to find a way out, and switch to `std::error_code` - but for now i'm resigned to waiting for the Networking TS to save us

Reply:
That all sounds perfectly sensible! Thanks.